Enable IValue to hold a PyObject (#30136)

zhaojuanmao opened the pull request:
Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/30136

This PR enables IValue to be able to hold a pure PyObject by adding a
new enum tag, a new jit_type to denote PyObject existance in IValue and
the JIT type system. We don't and not plan to expose this to user.

This is the basic piece that enable ivalue to be adopted broader like
making RRef always hold IValue, it might also simplify some compiler
logic

Test Plan: Imported from OSS

Differential Revision: D19502234

zhaojuanmao commented:
it is the same as #30136

zdevito reviewed (approved):
Stamping since other PR was approved.

zhaojuanmao commented:
@pytorchbot retest this please

zhaojuanmao committed:
Enable IValue to hold a PyObject (#32453)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32453

Pull Request resolved: https://github.com/pytorch/pytorch/pull/30136

This PR enables IValue to be able to hold a pure PyObject by adding a
new enum tag, a new jit_type to denote PyObject existance in IValue and
the JIT type system. We don't and not plan to expose this to user.

This is the basic piece that enable ivalue to be adopted broader like
making RRef always hold IValue, it might also simplify some compiler
logic

Test Plan: Imported from OSS

Differential Revision: D19502234

fbshipit-source-id: 31ad1df2e8624a9a09cf7c0c6482fd2c993884e3

zhaojuanmao commented:
somehow this PR can not run CI tests with error "Error response from daemon: Get https://308535385114.dkr.ecr.us-east-1.amazonaws.com/v2/pytorch/pytorch-linux-xenial-py3.6-clang7/manifests/07597f23-fa81-474c-8bef-5c8a91b50595: no basic auth credentials"

I tried to rebase and export it a few times and it just did not work, the CI test config was messed up somehow. 

I exported a new PR #32491 and now the tests are running normally. so closing this PR
